// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: teamcloud/v1/schedule_list_data_result.proto

package teamcloudv1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on ScheduleListDataResult with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ScheduleListDataResult) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ScheduleListDataResult with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ScheduleListDataResultMultiError, or nil if none found.
func (m *ScheduleListDataResult) ValidateAll() error {
	return m.validate(true)
}

func (m *ScheduleListDataResult) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Code

	for idx, item := range m.GetData() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ScheduleListDataResultValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ScheduleListDataResultValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ScheduleListDataResultValidationError{
					field:  fmt.Sprintf("Data[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if m.Status != nil {
		// no validation rules for Status
	}

	if m.Location != nil {
		// no validation rules for Location
	}

	if len(errors) > 0 {
		return ScheduleListDataResultMultiError(errors)
	}

	return nil
}

// ScheduleListDataResultMultiError is an error wrapping multiple validation
// errors returned by ScheduleListDataResult.ValidateAll() if the designated
// constraints aren't met.
type ScheduleListDataResultMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ScheduleListDataResultMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ScheduleListDataResultMultiError) AllErrors() []error { return m }

// ScheduleListDataResultValidationError is the validation error returned by
// ScheduleListDataResult.Validate if the designated constraints aren't met.
type ScheduleListDataResultValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ScheduleListDataResultValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ScheduleListDataResultValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ScheduleListDataResultValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ScheduleListDataResultValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ScheduleListDataResultValidationError) ErrorName() string {
	return "ScheduleListDataResultValidationError"
}

// Error satisfies the builtin error interface
func (e ScheduleListDataResultValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sScheduleListDataResult.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ScheduleListDataResultValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ScheduleListDataResultValidationError{}
